File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/notification/command/CustomEventCommandTest.java
package com.emarsys.mobileengage.notification.command;

import com.emarsys.mobileengage.MobileEngageInternal;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import java.util.HashMap;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

public class CustomEventCommandTest {

    public static final String EVENT_NAME = "eventName";

    private MobileEngageInternal mockMobileEngageInternal;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void setUp() {
        mockMobileEngageInternal = mock(MobileEngageInternal.class);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_mobileEngageInternal_mustNotBeNull() {
        new CustomEventCommand(null, "", new HashMap<String, String>());
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_eventName_mustNotBeNull() {
        new CustomEventCommand(mockMobileEngageInternal, null, new HashMap<String, String>());
    }

    @Test
    public void testRun_withEventAttributes() {
        HashMap<String, String> eventAttributes = new HashMap<>();
        eventAttributes.put("key", "value");

        Runnable customEventCommand = new CustomEventCommand(mockMobileEngageInternal, EVENT_NAME, eventAttributes);
        customEventCommand.run();

        verify(mockMobileEngageInternal).trackCustomEvent(EVENT_NAME, eventAttributes, null);
    }

    @Test
    public void testRun_withoutEventAttributes() {
        Runnable customEventCommand = new CustomEventCommand(mockMobileEngageInternal, EVENT_NAME, null);
        customEventCommand.run();

        verify(mockMobileEngageInternal).trackCustomEvent(EVENT_NAME, null, null);
    }

}

File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/notification/command/CustomEventCommandTest.kt
package com.emarsys.mobileengage.notification.command

import com.emarsys.mobileengage.event.EventServiceInternal
import io.kotest.assertions.throwables.shouldThrow
import io.mockk.mockk
import io.mockk.verify
import org.junit.Before
import org.junit.Test

class CustomEventCommandTest {
    companion object {
        private const val EVENT_NAME = "eventName"
    }

    private lateinit var mockEventServiceInternal: EventServiceInternal

    @Before
    fun setUp() {
        mockEventServiceInternal = mockk(relaxed = true)
    }

    @Test
    fun testConstructor_mockEventServiceInternal_mustNotBeNull() {
        shouldThrow<IllegalArgumentException> {
            CustomEventCommand(null, "", HashMap())
        }
    }

    @Test
    fun testConstructor_eventName_mustNotBeNull() {
        shouldThrow<IllegalArgumentException> {
            CustomEventCommand(mockEventServiceInternal, null, HashMap())
        }
    }

    @Test
    fun testRun_withEventAttributes() {
        val eventAttributes = HashMap<String, String>()
        eventAttributes["key"] = "value"
        val customEventCommand: Runnable =
            CustomEventCommand(mockEventServiceInternal, EVENT_NAME, eventAttributes)
        customEventCommand.run()
        verify { mockEventServiceInternal.trackCustomEventAsync(EVENT_NAME, eventAttributes, null) }
    }

    @Test
    fun testRun_withoutEventAttributes() {
        val customEventCommand: Runnable =
            CustomEventCommand(mockEventServiceInternal, EVENT_NAME, null)
        customEventCommand.run()
        verify { mockEventServiceInternal.trackCustomEventAsync(EVENT_NAME, null, null) }
    }
}